intellij/src/nl/tudelft/watchdog/intellij/logic/ui/listeners/EditorWindowListener.java

```diff
@@ -1,53 +1,76 @@
 import com.intellij.openapi.editor.Editor;
 import com.intellij.openapi.editor.event.EditorFactoryEvent;
 import com.intellij.openapi.editor.event.EditorFactoryListener;
+import com.intellij.openapi.editor.impl.DocumentMarkupModel;
+import com.intellij.openapi.editor.impl.MarkupModelImpl;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.Disposer;
+import nl.tudelft.watchdog.core.logic.event.TrackingEventManager;
+import org.jetbrains.annotations.NotNull;
 
 import java.util.HashMap;
 import java.util.Map;
 
-public class EditorWindowListener implements EditorFactoryListener {
+public class EditorWindowListener implements EditorFactoryListener, Disposable {
 
     private EditorFocusListener focusListener;
 
-    private Map<Editor, EditorListener> editorListenerMap = new HashMap<Editor, EditorListener>();
+    /**
+     * Every editor that is created is associated with a tree of disposables.
+     * The parent Disposable is registered with each associated listener to
+     * events of the editor. By using Disposable for this structure, we can
+     * properly cleanup all listeners once we dispose ourselves.
+     */
+    private Map<Editor, Disposable> editorDisposableTreeMap = new HashMap<>();
 
-    private String myProjectName;
+    private Project project;
+    private TrackingEventManager trackingEventManager;
 
-    public EditorWindowListener(String project) {
-        myProjectName = project;
+    EditorWindowListener(Project project, TrackingEventManager trackingEventManager) {
+        this.project = project;
+        this.trackingEventManager = trackingEventManager;
     }
 
     @Override
-    public void editorCreated(EditorFactoryEvent editorFactoryEvent) {
-        if(!editorBelongsToThisProject(editorFactoryEvent)) {
+    public void editorCreated(@NotNull EditorFactoryEvent editorFactoryEvent) {
+        if(editorDoesNotBelongToProject(editorFactoryEvent)) {
             return;
         }
 
         Editor editor = editorFactoryEvent.getEditor();
         focusListener = new EditorFocusListener(editor);
         editor.getContentComponent().addFocusListener(focusListener);
-        editorListenerMap.put(editor, new EditorListener(editor));
+
+        Disposable parentDisposable = Disposer.newDisposable();
+        Disposer.register(parentDisposable, new EditorListener(editor));
+        final MarkupModelImpl markupModel = (MarkupModelImpl) DocumentMarkupModel.forDocument(editor.getDocument(), this.project, true);
+        markupModel.addMarkupModelListener(parentDisposable, new IntelliJMarkupModelListener(trackingEventManager));
+
+        editorDisposableTreeMap.put(editor, parentDisposable);
     }
 
     @Override
-    public void editorReleased(EditorFactoryEvent editorFactoryEvent) {
-        if(!editorBelongsToThisProject(editorFactoryEvent)) {
+    public void editorReleased(@NotNull EditorFactoryEvent editorFactoryEvent) {
+        if (editorDoesNotBelongToProject(editorFactoryEvent)) {
```

Comment:
Why can there not be an NPE anymore?

Reply:
Because we now compare projects, rather than names. This means that if one of the arguments is `null`, it will compare that to an actual object.

Reply:
And project can never be null? (This comment should really be displayed a couple lines below)

Reply:
Per the API docs, our own `project` can not be null, but the project in `editorFactoryEvent` can be. Therefore, invoking `equals` on `project` is safe, but if we would flip it, it would not be safe.

Reply:
OK